Build for arm64 Macs

qvacua opened the pull request:
It compiles and runs without crashes. However, to compile, I had to add

```
DEPS_CMAKE_FLAGS += -DCMAKE_OSX_DEPLOYMENT_TARGET=11.00
```

to `local.mk` and do

```
make SDKROOT=`xcrun --show-sdk-path`
```

qvacua committed:
Use the latest luajit for arm64 Mac compatibility.

qvacua committed:
Do not set pagezero_size and image_base for LuaJIT >= 2.1.0-beta3

This is properly handled in LuaJIT now and setting causes "Malformed
Mach-o file" error when running the resulting binary on arm64 Macs.

jamessan commented on CMakeLists.txt:
```diff
@@ -352,7 +352,7 @@ if(CMAKE_C_COMPILER_ID STREQUAL "GNU" OR CMAKE_C_COMPILER_ID STREQUAL "Clang")
 endif()
 
-if(CMAKE_SYSTEM_NAME STREQUAL "Darwin" AND CMAKE_SIZEOF_VOID_P EQUAL 8)
+if(CMAKE_SYSTEM_NAME STREQUAL "Darwin" AND CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "x86_64" AND CMAKE_SIZEOF_VOID_P EQUAL 8 AND NOT PREFER_LUA)
```
It seems this isn't needed at all for LuaJIT >= 2.1.0 beta3.  That would be a better check than adding the architecture.

qvacua replied:
Changed the check to look for `LUAJIT_VERSION LESS "2.1.0-beta3".

jamessan commented on third-party/patches/libvterm-arm64-darwin.patch:
```diff
@@ -0,0 +1,10 @@
++++ b/bin/vterm-ctrl.c
+@@ -1,4 +1,4 @@
+-#define _XOPEN_SOURCE 500  /* strdup */
++#define _XOPEN_SOURCE 600  /* strdup */
```
This should be fixed upstream.  Cc @leonerd

qvacua replied:
Reverted the commit and created neovim/libvterm#6.

qvacua replied:
Updated the PR to use the newest revision of libvterm and adapted a line of code.

jamessan replied:
> and adapted a line of code.

That's an unreleased API change.  We shouldn't be pulling that in.  @leonerd, any chance of a maintenance release while the mainline stabilizes?

qvacua replied:
Ups, reverted the commit.

jamessan commented:
@leonerd Can you please release a new libvterm with the `_XOPEN_SOURCE` fix?  Also, it should be made unconditional.  That change is in no way specific to macOS.

clason commented:
@jamessan probably a candidate for backporting to `release-0.4`?  (same for #12676)

clason commented:
@qvacua I just tested this on non-ARM macOS and got 
```
/Library/Developer/CommandLineTools/usr/bin/make -C src
Makefile:319: *** missing: export MACOSX_DEPLOYMENT_TARGET=XX.YY.  Stop.
make[1]: *** [src/luajit] Error 2
```

jamessan commented:
> Maybe it would be a good idea to separate out the libvterm bump for easier backporting

Agreed.  Let's do that first, since it's a clean and obvious fix and then there can be follow up work for other issues specific to the macOS/arm problems.

justinmk commented on third-party/CMakeLists.txt:
```diff
@@ -161,8 +161,8 @@ set(UNIBILIUM_SHA256 29815283c654277ef77a3adcc8840db79ddbb20a0f0b0c8f648bd8cd49a
-set(LIBVTERM_URL https://github.com/neovim/libvterm/archive/65dbda3ed214f036ee799d18b2e693a833a0e591.tar.gz)
-set(LIBVTERM_SHA256 95d3c7e86336fbd40dfd7a0aa0a795320bb71bc957ea995ea0e549c96d20db3a)
+set(LIBVTERM_URL http://www.leonerd.org.uk/code/libvterm/libvterm-0.1.4.tar.gz)
+set(LIBVTERM_SHA256 bc70349e95559c667672fc8c55b9527d9db9ada0fb80a3beda533418d782d3dd)
```
yikes, no https? I really wish git-bzr did not mess up the hashes, that is the main problem with https://github.com/neovim/libvterm as a mirror .

clason replied:
No, `leonerd.org.uk` doesn't support https... (I checked for #12797)

What is the issue with the hashes? That they have to be recomputed manually?

justinmk replied:
> What is the issue with the hashes? That they have to be recomputed manually?

somehow git-bzr changes the hashes every time we update the mirror: https://github.com/neovim/bot-ci/blob/c799c16b003db67e353542ee7f1494731931dc6b/ci/sync-mirrors.sh#L46

not sure if we're using git-bzr wrong or if bzr just sucks.

justinmk replied:
actually it seems like that maybe isn't happening anymore ? if so then there's no reason to use leonerd.org.uk ...

clason replied:
except that the mirror isn't updated yet...

qvacua replied:
Ups, I did not notice the missing `s` in `http`. I'll update the PR to use https://github.com/neovim/libvterm/commit/54c03b21f763fa775a4c0643a9d8326342873179.

clason replied:
@qvacua the bump was already merged in #12797, by the way, so you can just rebase.

qvacua commented:
Neovim builds successfully and seems to work w/o problems on an arm64 Mac.

@clason : I bumped the LuaJIT version since there were recent commits which fixed the build on arm64 Macs. I am not sure about incompatibilities introduced by the bump though.

Regarding setting the deployment target: IMO, your suggestion is a good one. @justinmk , @jamessan: What is the officially supported minimum deployment target of Neovim? I guess 10.9, which @clason suggested, would also depend on the version of Xcode used. I think setting a default deployment target should be done in a different PR?

For the sake of completeness: AFAICS, there are usually three "sources" for deployment target when building on macOS

1. `libintl` is usually installed by `brew` with deployment target equal to the macOS you're currently using.
1. Setting `DEPS_CMAKE_FLAGS="-DCMAKE_OSX_DEPLOYMENT_TARGET=xyz" will set the deployment target of LuaJIT, but not other dependencies as it seems. Other dependencies simply do not have that information which can be checked by `otool` by looking for the `LC_VERSION_MIN_MACOSX` keyword.
1. Setting `MACOSX_DEPLOYMENT_TARGET=xyz` set the deployment target for other dependencies and the `nvim` binary itself.

It's not a big deal that the deployment targets of the dependencies and the `nvim` binary are different since no macOS specific APIs are used, I guess, but you most probably will get a warning at the end when building Neovim.

mikelxc commented:
Tried to build on MacBook Pro M1, LuaJIT failed to work out of box, an easy fix will be create a symlink for luajit in the bin folder and give it the name Lua.

However, even after that fix, still failed to build with error code
``` 
cd /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/runtime/pack/dist/opt/matchit && /opt/homebrew/Cellar/cmake/3.18.4/bin/cmake -E copy_directory /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/runtime/pack/dist/opt/matchit /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/runtime/pack/dist/opt/matchit
cd /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/runtime && /opt/homebrew/bin/luajit /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/scripts/genvimvim.lua /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/src/nvim /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/runtime/syntax/vim/generated.vim /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/funcs_data.mpack
cd /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/runtime && /opt/homebrew/Cellar/cmake/3.18.4/bin/cmake -E copy_directory /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/runtime/doc doc
cd /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/runtime/pack/dist/opt/vimball && /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/bin/nvim -u NONE -i NONE -e --headless -c helptags\ doc -c quit
cd /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/runtime/pack/dist/opt/matchit && /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/bin/nvim -u NONE -i NONE -e --headless -c helptags\ doc -c quit
/bin/sh: line 1: 73068 Killed: 9               /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/bin/nvim -u NONE -i NONE -e --headless -c helptags\ doc -c quit
/bin/sh: line 1: 73069 Killed: 9               /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/bin/nvim -u NONE -i NONE -e --headless -c helptags\ doc -c quit
make[2]: *** [runtime/pack/dist/opt/matchit/doc/tags] Error 137
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [runtime/pack/dist/opt/vimball/doc/tags] Error 137
cd /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/runtime && /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/bin/nvim -u NONE -i NONE -e --headless -c helptags\ ++t\ doc -c quit
/bin/sh: line 1: 73071 Killed: 9               /tmp/neovim-20201117-69850-1umpkrt/neovim-0.4.4/build/bin/nvim -u NONE -i NONE -e --headless -c helptags\ ++t\ doc -c quit
make[2]: *** [runtime/doc/tags] Error 137
make[1]: *** [runtime/CMakeFiles/runtime.dir/all] Error 2
make: *** [all] Error 2
```

clason commented:
You are trying to build an older released version that doesn't have the necessary patches, and using homebrew to boot. Can you please try building master (`git clone https://github.com/neovim/neovim.git`, `cd neovim`, `make`)?

mikelxc commented:
<img width="1440" alt="截屏2020-11-17 下午7 19 41" src="https://user-images.githubusercontent.com/20431963/99384076-daf70b80-2909-11eb-9341-1facdd4a3f3c.png">
<img width="1440" alt="截屏2020-11-17 下午7 20 02" src="https://user-images.githubusercontent.com/20431963/99384105-e64a3700-2909-11eb-9ee8-2e2a0a24f719.png">
Got the same error code for both brew install Neovim --HEAD and building from master

clason commented:
Thank you for testing. So it looks like neovim actually builds correctly, but the binary crashes or hangs(?) afterwards when generating help tags.

Could you try running `env VIMRUNTIME=./runtime ./build/bin/nvim` and see what happens?

mikelxc commented:
> Thank you for testing. So it looks like neovim actually builds correctly, but the binary crashes or hangs(?) afterwards when generating help tags.
> 
> Could you try running `env VIMRUNTIME=./runtime ./build/bin/nvim` and see what happens?

This is what I get 
```
[1]    42097 killed     env VIMRUNTIME=./runtime ./build/bin/nvim
```
I tried to delete the whole folder and clone it again and rebuilt it with 
`export MACOSX_DEPLOYMENT_TARGET=11.0  `
Still get the same error.

clason commented:
Can you get a stacktrace or something? That's not a lot to go on...

mikelxc commented:
Not 100% sure if I'm on the right track, but this is what I got
```
Undefined symbols for architecture arm64:
  "_custom_ui_autocmds_groups", referenced from:
      _apply_autocmds_group in fileio.c.o
  "_custom_ui_rpcevent_subscribed", referenced from:
      _nvim_subscribe in vim.c.o
  "_custom_ui_start", referenced from:
      _ui_builtin_start in ui.c.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

qvacua commented:
@mikelxc: At least on the DTK, the `master` does not run and is killed immediately similar to what you wrote. The reason is when the LuaJit library is linked `master` sets some parameters which do not work on arm64 Macs. This PR only sets those parameters only when appropriate. I've successfully compiled and run the branch of this [PR](https://github.com/qvacua/neovim/tree/build-arm64-mac) on the DTK.

qvacua commented:
BTW: Compilation succeeds only when `DEPS_CMAKE_FLAGS += -DCMAKE_OSX_DEPLOYMENT_TARGET=11.00` is set as in my first comment in this PR.

qvacua commented:
@mikelxc: The `_custom_ui_autocmds_groups` errors are specific to the `master` branch of my fork (for VimR). You should checkout the `build-arm64-mac` branch.

mikelxc commented:
It now builds without a problem after switching to the "build-arm64-mac" branch!

stefanpenner commented:
For anyone having issues, or simply lazy (like me), the following steps can likely be blindly followed:

```sh
git clone git@github.com:neovim/neovim
cd neovim
git remote add qvacua git@github.com:qvacua/neovim.git
echo "DEPS_CMAKE_FLAGS += -DCMAKE_OSX_DEPLOYMENT_TARGET=11.00" >> local.mk
make SDKROOT=`xcrun --show-sdk-path`
```

Resulting in a functioning nvim at `./build/bin/nvim`, and it is correctly built as arm64

```sh
~/s/n/neovim (build-arm64-mac)> file build/bin/nvim 
build/bin/nvim: Mach-O 64-bit executable arm64
```

menssen commented:
It's amazing how when somebody says "if you're lazy, just copy and paste this!" you tend to just do it, even when you know there's obviously a step missing. So if anybody else is _as lazy as I was_, here is the last comment, slightly corrected. (Either way, thanks @stefanpenner!)

```
git clone git@github.com:neovim/neovim
cd neovim
git remote add qvacua git@github.com:qvacua/neovim.git
git fetch qvacua
git checkout build-arm64-mac
echo "DEPS_CMAKE_FLAGS += -DCMAKE_OSX_DEPLOYMENT_TARGET=11.00" > local.mk
make SDKROOT=`xcrun --show-sdk-path`
```

jamessan committed:
third-party: Always set DEPLOYMENT_TARGET when building LuaJIT on macOS

LuaJIT build now requires specifying a deployment target, so use the
same baseline as our nightly builds.

Co-authored-by: Christian Clason <christian.clason@uni-due.de>

jamessan commented:
Looks like https://github.com/LuaJIT/LuaJIT/commit/ff34b48ddd6f2b3bdd26d6088662a214ba6b0288 brought out some issues in our test stack -- still not sure if it's in our code or busted.

Once we get a handle on those, we should be able to merge this PR.

mattnibs commented:
@jamessan is there any way I can help this along? Would be great to get this into master.

jamessan commented:
We need to figure out why some of our tests are failing like below with this PR and fix it.

```
  [  ERROR   ] 9 errors, listed below:
  [  ERROR   ] test/functional/plugin/lsp_spec.lua @ 265: LSP basic_init test should verify capabilities sent
  test/helpers.lua:73: exit code
  Expected objects to be the same.
  Passed in:
  (number) 101
  Expected:
  (number) 0
  ------------------------------------------------------------------------------
  $NVIM_LOG_FILE: Xtest-fake-lsp.log
  (last 999 lines)
  ERROR LSP: test/functional/fixtures/fake-lsp-server.lua:29: assert_eq failed: left == "\"shutdown\"", right == "\"test\""
  ERROR LSP: test/functional/fixtures/fake-lsp-server.lua:29: assert_eq failed: left == "{\
  ------------------------------------------------------------------------------
  
  
  stack traceback:
  	test/helpers.lua:73: in function 'eq'
  	test/functional/plugin/lsp_spec.lua:277: in function 'on_exit'
  	test/functional/plugin/lsp_spec.lua:109: in function 'test_rpc_server'
  	test/functional/plugin/lsp_spec.lua:269: in function <test/functional/plugin/lsp_spec.lua:265>
```

clason commented:
In particular, why the `capabilities` table is sorted differently in the two code paths that are checked there.

mattnibs committed:
Fix lsp tests breaking from new LuaJIT version

Apparently the new version of LuaJIT changed the consistency with which it
sorted table dictionaries. IIRC lua sorts dictionary keys by memory address, so
it would appear that the reasons tests were previously passing was because of
a differentiation in the implementation of the lua runtime.

Ensure that array fields in the lsp protocol tables are consistently created,
by using ipair when generating arrays for completionItemKind and
symbolItemKind.

For CodeActionKind, the current implementation includes both the keys and the
values in the array. This is incorrect. Ensure that only the values are
included in the array and sort them for consistency.

mattnibs commented:
I believe I have a fix for this here: https://github.com/qvacua/neovim/pull/1

(Wasn't sure the appropriate was of including this with this pr).

TC72 commented:
> It's amazing how when somebody says "if you're lazy, just copy and paste this!" you tend to just do it, even when you know there's obviously a step missing. So if anybody else is _as lazy as I was_, here is the last comment, slightly corrected. (Either way, thanks @stefanpenner!)
> 
> ```
> git clone git@github.com:neovim/neovim
> cd neovim
> git remote add qvacua git@github.com:qvacua/neovim.git
> git fetch qvacua
> git checkout build-arm64-mac
> echo "DEPS_CMAKE_FLAGS += -DCMAKE_OSX_DEPLOYMENT_TARGET=11.00" > local.mk
> make SDKROOT=`xcrun --show-sdk-path`
> ```



> It's amazing how when somebody says "if you're lazy, just copy and paste this!" you tend to just do it, even when you know there's obviously a step missing. So if anybody else is _as lazy as I was_, here is the last comment, slightly corrected. (Either way, thanks @stefanpenner!)
> 
> ```
> git clone git@github.com:neovim/neovim
> cd neovim
> git remote add qvacua git@github.com:qvacua/neovim.git
> git fetch qvacua
> git checkout build-arm64-mac
> echo "DEPS_CMAKE_FLAGS += -DCMAKE_OSX_DEPLOYMENT_TARGET=11.00" > local.mk
> make SDKROOT=`xcrun --show-sdk-path`
> ```

When I follow this I get:
FAILED: src/nvim/CMakeFiles/nvim.dir/tui/tui.c.o

clason commented:
@TC72 make sure to `make distclean` beforehand. Also, this PR is now merged, so if you update neovim to `HEAD`, you should only need the last two lines. If these don't work for you, please open a new issue with full information to reproduce this (especially your system).

Keating950 commented:
@clason, for what it's worth, I didn't even need the last two lines to build successfully on my M1 mini.

TC72 commented:
> @clason, for what it's worth, I didn't even need the last two lines to build successfully on my M1 mini.

Is there any chance you could walk through exactly how you did it. I've tried everything and still can't get it to build. There's obviously something wrong with my setup if it's working for others.

yingzhu146 commented:
```
ninja  -C build
ninja: Entering directory `build'
[301/321] Building C object src/nvim/CMakeFiles/nvim.dir/tui/tui.c.o
FAILED: src/nvim/CMakeFiles/nvim.dir/tui/tui.c.o 
/Library/Developer/CommandLineTools/usr/bin/cc -DINCLUDE_GENERATED_DECLARATIONS -DMIN_LOG_LEVEL=1 -DNVIM_MSGPACK_HAS_FLOAT32 -D_GNU_SOURCE -Dnvim_EXPORTS -Iconfig -I../src -Isrc/nvim/auto -Iinclude -I../.deps/usr/include/luajit-2.1 -isystem ../.deps/usr/include -isystem /opt/local/include -isystem /Library/Frameworks/R.framework/Headers -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -g -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -MD -MT src/nvim/CMakeFiles/nvim.dir/tui/tui.c.o -MF src/nvim/CMakeFiles/nvim.dir/tui/tui.c.o.d -o src/nvim/CMakeFiles/nvim.dir/tui/tui.c.o -c ../src/nvim/tui/tui.c
../src/nvim/tui/tui.c:552:7: error: no member named 'i' in 'unibi_var_t'
      UNIBI_SET_NUM_VAR(data->params[0], standout);
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Same here @TC72  - even after `make distclean`, and after deleting and pulling from master from fresh install

TC72 commented:
> ```
> ninja  -C build
> ninja: Entering directory `build'
> [301/321] Building C object src/nvim/CMakeFiles/nvim.dir/tui/tui.c.o
> FAILED: src/nvim/CMakeFiles/nvim.dir/tui/tui.c.o 
> /Library/Developer/CommandLineTools/usr/bin/cc -DINCLUDE_GENERATED_DECLARATIONS -DMIN_LOG_LEVEL=1 -DNVIM_MSGPACK_HAS_FLOAT32 -D_GNU_SOURCE -Dnvim_EXPORTS -Iconfig -I../src -Isrc/nvim/auto -Iinclude -I../.deps/usr/include/luajit-2.1 -isystem ../.deps/usr/include -isystem /opt/local/include -isystem /Library/Frameworks/R.framework/Headers -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -g -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -MD -MT src/nvim/CMakeFiles/nvim.dir/tui/tui.c.o -MF src/nvim/CMakeFiles/nvim.dir/tui/tui.c.o.d -o src/nvim/CMakeFiles/nvim.dir/tui/tui.c.o -c ../src/nvim/tui/tui.c
> ../src/nvim/tui/tui.c:552:7: error: no member named 'i' in 'unibi_var_t'
>       UNIBI_SET_NUM_VAR(data->params[0], standout);
>       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> Same here @TC72 - even after `make distclean`, and after deleting and pulling from master from fresh install

There is a patch which fixes everything, I'm now able to build for a native version. Have a look at this thread:
https://github.com/neovim/neovim/issues/13533#issuecomment-745772611

BUT, checking there now there may be a new issue.

andir commented on third-party/CMakeLists.txt:
```diff
@@ -147,8 +147,8 @@ set(MSGPACK_URL https://github.com/msgpack/msgpack-c/releases/download/cpp-3.0.0
 set(MSGPACK_SHA256 bfbb71b7c02f806393bc3cbc491b40523b89e64f83860c58e3e54af47de176e4)
 
 # https://github.com/LuaJIT/LuaJIT/tree/v2.1
```
Mhm, I guess this should have been updated to the rev this now points at?
